spd/scripts/compare_models.py

```diff
@@ -0,0 +1,393 @@
+            ref_U = reference_components.U
+            ref_V = reference_components.V
+
+            # Filter out components that aren't active enough in the current model
+            C_alive = sum(activation_densities[layer_name] > self.density_threshold)
+            if C_alive == 0:
+                logger.warning(
+                    f"No components are active enough in {layer_name} for density threshold {self.density_threshold}. Skipping."
+                )
+                continue
+
+            current_U_alive = current_U[activation_densities[layer_name] > self.density_threshold]
+            current_V_alive = current_V[
+                :, activation_densities[layer_name] > self.density_threshold
+            ]
+
+            # Compute rank-one matrices: V @ U for each component
+            current_rank_one = einops.einsum(
+                current_V_alive,
+                current_U_alive,
+                "d_in C_alive, C_alive d_out -> C_alive d_in d_out",
+            )
+            ref_rank_one = einops.einsum(ref_V, ref_U, "d_in C, C d_out -> C d_in d_out")
+
+            # Compute cosine similarities between all pairs
+            current_flat = current_rank_one.reshape(int(C_alive.item()), -1)
+            ref_flat = ref_rank_one.reshape(C, -1)
+
+            current_norm = F.normalize(current_flat, p=2, dim=1)
+            ref_norm = F.normalize(ref_flat, p=2, dim=1)
+
+            cosine_sim_matrix = einops.einsum(
+                current_norm, ref_norm, "C_alive d_in_d_out, C_ref d_in_d_out -> C_alive C_ref"
+            )
+            cosine_sim_matrix = cosine_sim_matrix.abs()
+
+            max_similarities = cosine_sim_matrix.max(dim=1).values
+            similarities[f"mean_max_abs_cosine_sim/{layer_name}"] = max_similarities.mean().item()
+            similarities[f"max_abs_cosine_sim_std/{layer_name}"] = max_similarities.std().item()
+            similarities[f"max_abs_cosine_sim_min/{layer_name}"] = max_similarities.min().item()
+            similarities[f"max_abs_cosine_sim_max/{layer_name}"] = max_similarities.max().item()
+
+        metric_names = [
+            "mean_max_abs_cosine_sim",
+            "max_abs_cosine_sim_std",
+            "max_abs_cosine_sim_min",
+            "max_abs_cosine_sim_max",
+        ]
+
+        for metric_name in metric_names:
+            values = [
+                similarities[f"{metric_name}/{layer_name}"]
+                for layer_name in self.current_model.components
+                if f"{metric_name}/{layer_name}" in similarities
+            ]
+            if values:
+                similarities[f"{metric_name}/all_layers"] = sum(values) / len(values)
+
+        return similarities
+
+    def run_comparison(
+        self, eval_iterator: Iterator[Any], n_steps: int | None = None
+    ) -> dict[str, float]:
+        """Run the full comparison pipeline."""
+        if n_steps is None:
+            n_steps = self.comparison_config.get("n_eval_steps", 5)
```

Comment:
this default pattern is a bit scary, but maybe it's OK if you think it is useful for this specific script.